### ISSUE TITLE
Validation with custom closure broken

### ISSUE THREAD

**Issue by WTFDuck** [cakephp/cakephp#17440]
### Description

I attempted to validate that a user has chosen an item in a select box, this validation gives predicted result and displays the validation message when field is empty:

` $validator->allowEmptyString('somefield', 'This is required', FALSE );`

Using a closure as below should give same result as above on submitting an empty field, but it does not. It has some unpredictable behaviors, like no validation error on that field whatsoever, as if the rule were ignored.

`$validator->allowEmptyString('somefield', 'This is required', function ($context) {
                                 return FALSE;
                            }); `

 If the closure above was set to return TRUE if gives the default validation error of "Please select an item from the list" instead of my inputted message.  This is backwards behavior, and my message is missing.

To sum up, any logic I put in the closure seems to give the same goofiness. All validation types, notEmptyString() allowEmptyString() and even add() return the same oddity when using closures/anonymous functions.  I haven't tested any other validation types, however.

This was as seen on Cake 5.0.2, but I also experienced this on Cake 4.5 when I tested this evening.  

Thanks much for all you do!

::: Edited for clarity :::

### CakePHP Version

5.0.2

### PHP Version

8.2.12

**Comment by markstory** [comment]
Thanks for opening an issue. I've added a test based on your example, and it is returning what I was thinking would happen and what you described as wanting. What am I missing?

**Comment by WTFDuck** [comment]
Hi Mark!  Thank you for the quick response and HAPPY THANKSGIVING wherever you are at! 

As far as the question, well, its not clear to me, but I what I "want" is for a closure to work properly in a validation.  They're not working, they do odd things, and don't work even when code is copied from the manual on a fresh install.  Maybe I have a borked installation, or maybe my php install is doing something peculiar.  But when I saw that cakephp 4.5 did the same thing I was rather surprised that nobody brought this to attention before.  Custom closures may be relatively unused feature that somehow this slipped through the cracks?

So you said it returned what you were thinking... what were you thinking?  What did it return for you?  I suppose you got similar results to what I did.  But you asked what you're missing, so maybe it worked as per manual?  Not sure, bit ambiguous.

No matter, its still broken here, lol!

**Comment by markstory** [comment]
> HAPPY THANKSGIVING wherever you are at!

Happy Thanksgiving to you as well.

> So you said it returned what you were thinking... what were you thinking? What did it return for you? I suppose you got similar results to what I did. But you asked what you're missing, so maybe it worked as per manual? Not sure, bit ambiguous.

I'm expecting the behavior to match what we've got documented, or how it has historically if the behavior had tests, as tests more than our documentation are our long term contract with developers as it is how we document all the edgecases and fixes that have been made.

I see know how what the documentation says which is:

>      * @param \Closure|string|bool $when Indicates when the field is allowed to be empty
>     * Valid values are true, false, 'create', 'update'. If a Closure is passed then
>     * the field will allowed to be empty only when the callback returns true.

And that the current behavior could be seen to differ. But I don't think they do.

```php
$validator = new Validator();
$validator->allowEmptyString(
    'title',
    'very required',
    function ($context) {
        return $context['data']['otherField'] === true;
    }
);

$data = [
    'title' => '',
    'otherField' => false,
];
$result = $validator->validate($data);
$this->assertNotEmpty($result);
$this->assertEquals(['_empty' => 'very required'], $result['title']);
```

Looks correct to me. Because `otherField` is `false`, the field is not allowed to be empty. We also have a test for the opposite case

```php
$data = [
    'title' => '',
    'otherField' => true,
];
$this->assertEmpty($validator->validate($data));
```

This also makes sense to me as `otherField` is `true` so the field is allowed to be empty. 

> what I "want" is for a closure to work properly in a validation. They're not working, they do odd things, and don't work even when code is copied from the manual on a fresh install. Maybe I have a borked installation, or maybe my php install is doing something peculiar. 

You should be able to run these examples locally (minus the assert methods). If you dump out the results you should see similar results. If you don't - that would be peculiar.

> what I "want" is for a closure to work properly in a validation. 

If you were to formalize this further what would the expected behavior be for the closure returning true and false?

**Comment by WTFDuck** [comment]
THANK YOU for being so thorough!   I am floored... Running that test you provided it does indeed function as it should, and your test and example is nearly verbatim of what I was trying to perform before I reported anything.  

HOWEVER... when I apply this allowEmptyString() validation rule with a closure function in my ModelTable, then receiving the $data from a submitted form, the form still does't invalidate this field as it should.  I still get the same goofiness of default messages on my select boxes and broken results that I described initially.  But, again, only when there is a closure function involved.

Seems I may have pointed the finger the direction, I was so sure validation was the problem,  but it seems the problem may lie elsewhere.  Gotta do more digging.  I just don't understand why I can copy from the quick start manual verbatim and that doesn't work either.  ¯\_(ツ)_/¯

**Comment by WTFDuck** [comment]
Using a form to submit into the test validation with all the validation in the controller gives me results that I was expecting, and works as per manual! YAY!  THANK YOU!

But if I add that EXACTLY SAME RULE to my validationDefault function in my ModelTable.... the results are all kinds of borked.  Backwards from anticipated, missing my custom error messages etc.  Really strange behaviors.

Validation in and of itself is otherwise fine, there appears to be some kind of processing of the rule when created in the ModelTable when it contains a closure function.

**Comment by markstory** [comment]
> Validation in and of itself is otherwise fine, there appears to be some kind of processing of the rule when created in the ModelTable when it contains a closure function.

Interesting, as `validationDefault` gets passed a `Validator` much like the standalone usage we've been doing. Do you have an example of the validation method you're trying to build and the data you're validating?

**Comment by WTFDuck** [comment]
Its a motorcycle parts website, need to choose make of manufacturer when adding a listing, if make is not in select box, then must enter a "make_new" in the following input box. Simple enough.

ModelTable formerly contain(ed) this.  It is also the same rule that is working reliably from within the controller currently. It seems happiest in my use with a !empty() in the return:

```php
 public function validationDefault(Validator $validator): Validator
    {
            $validator->allowEmptyString(
                'make',
                'Need to choose at least one',
                function ($context) {
                    return !empty($context['data']['make_new']) === true;
                }
            );
     }
``` 

And in the form: 

```php
echo $this->Form->create($listing, ['url' => ['controller' => 'Admin', 'action' => 'listingAdd']]); 
echo $this->Form->control('make', ['options' => $makes, 'empty' => "Please Choose"]); 
echo $this->Form->control('make_new'); 
echo $this->Form->button(__('Save Item')); 
echo $this->Form->end(); 
``` 

The available makes in the select:
```php
$makes = array('Ducati'=>'Ducati', 'Triumph' => 'Triumph', 'Honda' => 'Honda');
``` 
An example is to input a make_new of Yamaha in the input box (or don't), and things get really silly. 

The default "Please select an item from the list" message comes up on make when both boxes empty, SAME when make is empty, and make_new is filled.

I attempted to tweak the return rule in the closure function, and managed to get it to invalidate when make is NOT selected, but make_new was filled.  Backwards behavior which I couldn't break it from.

But again, I put the same rules in the controller and they are working just fine and were put into production over the weekend.

It would be a wonderful world if we didn't need these sort of things, and average users could clean their own data input, meanwhile we have to spend half or our development time on things like this.  LoL!

**Comment by markstory** [comment]
Interesting that you're seeing different behavior in the model and controller defined validation. `Validator` objects aren't aware of the context that they were defined in. I'm assuming the 'please select an item from the list' error message is defined in a different validator method? Does your validator behave correctly without those additional rules?

**Comment by othercorey** [comment]
@WTFDuck Have you been able to try this out?

**Comment by gagnav** [comment]
It seems I have this kind of issue as well, the NotEmptyString closure function for conditional validation does not behave properly.
```php
 $validator
            ->decimal('price')
            ->requirePresence('price', 'create', '"Price" field should be present in dataset.')
            ->notEmptyString('price', '"Price" field cannot be left empty.', function($context){
                return $context['data']['pricing_type'] === 'normal';
            });
```
So I have a price field that can be either "normal" or "free", for some strange reason this gives me a warning in view 'Undefined array key "pricing_type" [in /var/www/html/src/Model/Table/ProductsTable.php, line 117]' while just loading the view so the form is not submitted and I get this warning just by loading the page. So essentially the $context['data'] is not populated with any data and it should not be because the form is not submitted, so there is no POST request and patchEntity is not executed in controller.
To make things even weirder when I move the same closure into requirePresence it just works fine and as expected so the following code works without issues.
```php
$validator
          ->decimal('price')
          ->requirePresence('price', function($context){
                return $context['data']['pricing_type'] === 'normal';
            }, '"Price" field should be present in dataset.')
            ->notEmptyString('price', '"Price" field cannot be left empty.');
```
            
I guess I will just rewrite the logic a bit to use requirePresence instead of NotEmptyString, but the behavior is a bit weird.

PHP 8.2.14
CakePHP 4.4.17
OS Ubuntu 22.04.3 LTS

**Comment by ADmad** [comment]
> while just loading the view so the form...

That's because the `FormHelper` calls `EntityContext::isRequired()` to check whether a field is allowed to be empty, to decide whether the `required` attribute should be set for the input.

https://github.com/cakephp/cakephp/blob/5.x/src/View/Form/EntityContext.php#L496

**Comment by gagnav** [comment]
Thank you very much for your response.
I was relying on the book
[](https://book.cakephp.org/4/en/core-libraries/validation.html#conditional-validation)
After your response I have modified the code to the following and it works. Essentially, we just need to add a default return value to the closure function and check values set by user for conditional check.
```php
 $validator
    ->decimal('price')
    ->requirePresence('price', 'create', '"Price" field should be present in dataset.')
    ->notEmptyString('price', '"Price" field cannot be left empty.', function($context){
                if(isset($context['data']['pricing_type'])){
                    return $context['data']['pricing_type'] === 'normal';
                }
                return true;
            });
```

**Comment by WTFDuck** [comment]
Thanks guys, I've been away for quite a while.  Had a lot going on around here, mostly spent time under classic VWs getting dirty.  I appreciate everyone addressing this.  Thank you, again!
